### DISCUSSION TITLE
helm已部署v6d集群之后

### DISCUSSION THREAD

**Discussion by nanyoullm** [v6d-io/v6d#767] [Q&A] [answered]
k8s两台机器（master+node），已完成helm方式部署v6d
![image](https://user-images.githubusercontent.com/13297712/170972888-617931e0-71c2-4f48-83f0-df7a3d482b81.png)
接下来的文档未看懂，不是已经在k8s部署了v6d么，咋还需要启动一个守护服务呢，是在master启动吗
![image](https://user-images.githubusercontent.com/13297712/170973224-d2b016fc-7924-49fb-b5b0-5ac68df16953.png)
尝试在master执行`python -m vineyard`，报错如下：
![image](https://user-images.githubusercontent.com/13297712/170973382-f2003a37-0d28-4bf5-bec5-eb5de9fc7079.png)
是什么问题呢

**Comment by nanyoullm** [comment]
实在是看不懂官方的文档，国内做的东西也没汉化，有熟悉的同学指导下helm部署后怎么在客户端使用吗

**Comment by sighingnow** [comment, marked as answer]
Hi @nanyoullm,

这个文档确实有点混淆。具体是这样的，Vineyard可以在本地单机部署，也可以在Kubernetes上部署（使用Helm），这是两种不同的方式。

- 本地部署
   - 运行 `python -m vineyard`，之后在 `/var/run/vineyard.sock` 处有一个可以连接的UNIX-domain socket
   - 这个socket的路径可以使用 `--socket`参数来指定，例如 `python -m vineyard --socket /tmp/vineyard.sock`

- 在Kubernetes上使用Helm部署
   - 在Kubernetes上，Vineyard 的 Pod通常是与其他一样部署在Kubernetes上的Pod通过一个Volume来共享这个socket
   - 部署在Kubernetes上时，因为Vineyard的pod运行在K8s的worker node上，因此在Master上没办法建立IPC连接。

希望能解答你关于Vineyard的疑惑～

**Reply by nanyoullm** [reply]
@sighingnow  感谢 hetao 大神的解答，确实清晰了很多。
还有个问题是：在 k8s 部署的方式中，client 处理好的数据怎么和 vineyard 进行交互呢？
我看 pythonAPI 中还可以通过 TCP 连接 vineyard，所以我应该使用这种方式吗
![image](https://user-images.githubusercontent.com/13297712/171077193-15c9b4d3-6b17-48e6-822f-0db53d347a4c.png)

尝试了连接 endpoint 的方式，但是在 `helm install vineyard vineyard/vineyard` 后，通过 `kubectl get endpoint` 后显示
![image](https://user-images.githubusercontent.com/13297712/171078381-8bdbdd4b-50b1-42af-9470-0ed098f19356.png)
如果我说的方式是没错的话，应该是连哪个 IP 和 PORT 呢（试了几个都貌似不行）

`helm install vineyard vineyard/vineyard` 后根据提示执行 `kubectl -n default describe service vineyard-rpc`，如下图
![image](https://user-images.githubusercontent.com/13297712/171079448-6d3863d0-7aac-4702-ba53-66a56b9f8204.png)

**Reply by sighingnow** [reply]
> 那在 k8s 部署的方式中，client 处理好的数据怎么和 vineyard 进行交互呢

Kubernetes部署的话，我们一般需要用户的workload和vineyard是部署在一起的，例如在GraphScope中，graphscope的worker和vineyard的container会在同一个pod里面。

> 我看 pythonAPI 中还可以通过 TCP 连接 vineyard

这个一般是用来获取vineyard中的一些metadata。

> 如果我说的方式是没错的话，应该是连哪个 IP 和 PORT 呢（试了几个都貌似不行）

我double check一下helm charts里面暴露的rpc endpoint。

**Reply by nanyoullm** [reply]
所以是否可以这样理解呢：如果想交互式学习 vineyard 的话，还是本地部署更方便实现交互；基于 Helm 部署 k8s 的方式无法进行交互，需要和其他 workload （比如数据处理的上下游 container？）绑定在 pod 里运行。

**Reply by sighingnow** [reply]
> 所以是否可以这样理解呢：如果想交互式学习 vineyard 的话，还是本地部署更方便实现交互；

是的。

> 基于 Helm 部署 k8s 的方式无法进行交互，需要和其他 workload （比如数据处理的上下游 container？）绑定在 pod 里运行。

这个交互式连接现在阶段确实有点麻烦

**Comment by sighingnow** [comment]
具体到截图的中的问题，看起来是因为etcd使用的端口和系统中其他程序冲突了（Vineyard依赖etcd），你可以通过

```
python3 -m vineyard --etcd_endpoint=http://127.0.0.1:2479
```

来指定一个空闲的端口使用。

**Comment by sighingnow** [comment]
这个discussion里面提到的问题，见issue：https://github.com/v6d-io/v6d/issues/771

文档会更新。

**Comment by sighingnow** [comment]
> 尝试在master执行python -m vineyard，报错如下：
> ![image](https://user-images.githubusercontent.com/7144772/171079968-2a09e4be-c052-4257-813c-33b80b004a5c.png)
>
> 是什么问题呢

这个问题已经在pull request https://github.com/v6d-io/v6d/pull/770 中fix。

**Reply by nanyoullm** [reply]
👍👍👍
